### PR TITLE
Add ability to ingest a PDF to be split apart into paged content.

### PR DESCRIPTION
**Pull request by jordandukart**

### CODE REVIEW

**Comment by willtp87** [comment]
Everywhere there is 'book' in a description or function name it should rather say something like 'paged object'.

File: islandora_paged_content.module [hunk 1]
```diff
@@ -238,3 +238,14 @@ function swaptable_value($element, $input, array $form_state) {
   }
   return array();
 }
+
+/**
+ * Implements hook_islandora_object_ingested().
+ */
+function islandora_paged_content_islandora_object_ingested(AbstractObject $object) {
+  // Only do this check on paged content that have PDFs.
+  if (array_intersect_key(array('islandora:bookCModel', 'islandora:newspaperIssueCModel'), $object->models) && isset($object['PDF'])) {
```
Comment (willtp87) [inline islandora_paged_content.module]: I know we are doing something like this already, but we really shouldn't.  We should create a registry hook for page content models. Also maybe manuscripts should be included as in https://github.com/Islandora/islandora_paged_content/blob/0db9de2906219186d1922b065b1be165dafed2f2/includes/utilities.inc#L298 but it shouldn't be as manuscript is a DGI module which just reinforces the hook notion.

File: includes/batch.inc [hunk 2]
```diff
@@ -425,3 +425,107 @@ function islandora_paged_content_delete_parent_object_operation($pid) {
...
+function islandora_paged_content_create_book_from_pdf_batch(AbstractObject $paged_content, array $pages, array $options) {
+  return array(
+    'operations' => array(
+      array('islandora_paged_content_create_page_from_pdf_batch_operation', array(
+        $pages, $options, $paged_content->id)),
+    ),
+    'finished' => 'islandora_paged_content_batch_finished',
+    'title' => t('Creating pages from @label ...', array('@label' => $paged_content->label)),
+    'init_message' => t('Preparing to extract pages from @label ...', array('@label' => $paged_content->label)),
+    'progress_message' => t('Time elapsed: @elapsed <br/>Estimated time remaining @estimate.'),
```
Comment (willtp87) [inline includes/batch.inc]: Should maybe have a colon at 'Estimated time remaining:'

File: includes/batch.inc [hunk 3]
```diff
@@ -425,3 +425,107 @@ function islandora_paged_content_delete_parent_object_operation($pid) {
...
+    'error_message' => t('An error has occurred.'),
+    'file' => drupal_get_path('module', 'islandora_paged_content') . '/includes/batch.inc',
+  );
+}
+
+/**
+ * Batch operation for converting a page into a PDF.
```
Comment (willtp87) [inline includes/batch.inc]: Other way around?

File: includes/batch.inc [hunk 4]
```diff
@@ -425,3 +425,107 @@ function islandora_paged_content_delete_parent_object_operation($pid) {
...
+ *   images.
+ *   -resolution: A string specifying the DPI of the to be generated image.
+ *   -language: A string specifying the language of the OCR text.
+ *   -extract_text: An integer specifying whether to extract text from the PDF
+ *   page or to let tesseract or pdftext work their magic for us.
+ *   -model: A string specifying the content model of the parent.
+ *   -pdf_file: A Drupal file object of the uploaded PDF to be processed.
+ *   -parent: A string specifying the parent pid of the to be ingested pages.
+ * @param string $pid
+ *   The PID of the parent pid.
```
Comment (willtp87) [inline includes/batch.inc]: pid of a pid eh

File: includes/batch.inc [hunk 5]
```diff
@@ -425,3 +425,107 @@ function islandora_paged_content_delete_parent_object_operation($pid) {
...
+ *   An array of options to pass into ghostscript and pdftotext containing:
+ *   -image_format: A string specifying the device to be used when generating
+ *   images.
+ *   -resolution: A string specifying the DPI of the to be generated image.
+ *   -language: A string specifying the language of the OCR text.
+ *   -extract_text: An integer specifying whether to extract text from the PDF
+ *   page or to let tesseract or pdftext work their magic for us.
+ *   -model: A string specifying the content model of the parent.
+ *   -pdf_file: A Drupal file object of the uploaded PDF to be processed.
+ *   -parent: A string specifying the parent pid of the to be ingested pages.
```
Comment (willtp87) [inline includes/batch.inc:483]: Of the what?

File: includes/upload_pdf.form.inc [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  if ($form_state['values']['pdf_file']) {
+    module_load_include('inc', 'islandora_paged_content', 'includes/utilities');
+    $pdf_file = $form_state['values']['pdf_file'];
+    $file_obj = file_load($pdf_file);
+    $num_pages = islandora_paged_content_length_of_pdf($file_obj->uri);
+    if ($num_pages) {
+      $form_state['number_of_pages'] = $num_pages;
+    }
+    else {
+      form_error($form['pdf_file'], t('Pdfino was unable to extract required information from the supplied PDF.'));
```
Comment (willtp87) [inline includes/upload_pdf.form.inc]: Probably shouldn't mention a util in an end user message.

File: includes/utilities.inc [hunk 7]
```diff
@@ -987,3 +990,68 @@ function islandora_paged_content_extract_tiff_from_pdf($uri, $offset, $device, $
...
+ *   An AbstractObject representing an object within Fedora.
+ *
+ * @param object $file_obj
+ *   The PDF file object that is being added.
+ */
+function islandora_paged_content_add_pdf_to_object(AbstractObject $object, $file_obj) {
+  module_load_include('inc', 'islandora_paged_content', 'includes/db');
+  if (!isset($object['PDF'])) {
+    $pdf = $object->constructDatastream('PDF', 'M');
+    $object->ingestDatastream($pdf);
```
Comment (willtp87) [inline includes/utilities.inc]: Should move this to the bottom of the function so it won't create extra versions if the object is already ingested in the future.

File: includes/batch.inc (hunk 5 above)
Comment (jordandukart) [inline includes/batch.inc:483]: Think this is the clear? It's the parent of the pages that are being ingested.

Comment (willtp87) [inline includes/batch.inc:483]: Yeah, my bad structure got me.

**Pull request merged**